docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,604 @@
+- Reformat a log line to be more readable.
+
+However those use cases are out of scope:
+
+- What's the latency xth quantile over a year, we don't expect time range to be higher than couple of day max.
+- Replace metrics instrumentation with logs. We want to ease new analysis but we don't want to push users to use this instead of instrumenting their application with Prometheus or any other metric client.
+- Big data analysis, this whole new feature is mainly for ad-hoc use cases.
+
+## LogQL Parsers
+
+A parser operator can only be applied to __log__ streams (a set of entries attached to a given set of label pairs) and will transform (or extract) them into __field__ streams.
+
+As quick overview let's consider stream below:
+
+```json
+{
+  "stream": {
+   "job": "loki-dev/ingester",
+   "cluster": "us-central1",
+   "namespace" : "loki-dev",
+  },
+  "values": [
+    [
+      "1584114643856675277",
+      "level=info ts=2020-03-13T15:50:43.856555 caller=metrics.go"
+    ],
+    [
+      "1584114643856675278",
+      "level=warn ts=2020-03-13T15:50:43.856556 caller=metrics.go"
+    ]
+  ]
+}
+```
+
+As you can see the log is using the popular logfmt so we could use the `| logfmt` parsers to transform those entries into the fields entries like below:
+
+```json
+{
+  "stream": {
+   "job": "loki-dev/ingester",
+   "cluster": "us-central1",
+   "namespace" : "loki-dev",
+  },
+  "values": [
+    [
+      "1584114643856675277",
+      {
+          "level":"info",
+          "ts": "2020-03-13T15:50:43.856555",
+          "caller": "metrics.go"
+      }
+    ],
+   [
+      "1584114643856675278",
+      {
+          "level":"warn",
+          "ts": "2020-03-13T15:50:43.856556",
+          "caller": "metrics.go"
+      }
+    ]
+  ]
+}
+```
+
+Field names are declared by each parsers, some parser will allow implicit fields extraction for __brevity__ like the `json` and `logfmt` parsers.
+
+Since field name can conflicts with labels during aggregation we will support for each parser the ability to rename a field. However in case of conflict fields will take precedence over labels.
+
+> Implicit fields extraction could means trouble in term of performance. However in most cases we will be able to infer all used fields from the full query and so limit the extraction per line.
+
+To cover a wide range of structured and unstructured logs, we will start with those four parsers: `glob`, `logfmt`, `json` and `regexp`.
+
+### Glob
+
+The glob parser (as described in previous design) `{job="prod/nginx"} | glob "*<client_ip> - - [ * ] \"* *<path> "` could be used to extract fields from nginx logs.
+
+The log entry below:
+
+```log
+127.0.0.1 - - [19/Jun/2012:09:16:22 +0100] "GET /GO.jpg HTTP/1.1" 499 0 "http://domain.com/htm_data/7/1206/758536.html" "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; .NET CLR 1.1.4322; .NET CLR 2.0.50727; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; SE 2.X MetaSr 1.0)"
+```
+
+would be transformed into those fields:
```

Comment:
Are field key/value pairs only strings? Obviously keys are strings, but can values be something else, e.g. arrays?

We are doing similar to Canonical Log Lines ([ref 1](https://brandur.org/canonical-log-lines), [ref 2](https://stripe.com/blog/canonical-log-lines)), but with all messages for a request nested under a `messages` key, in an array.  So basically every http request is one log line.

It looks like this:

```json
{
  "request-id" : "2d94738b-0324-a1fc-2916-644716f3ea36",
  "ip-address" : "127.0.0.1",
  "user-name" : "Jane Doe",
  "http-method" : "get",
  "http-response" : "200",
  "duration" : 125,
  "service" : "authentication",
  "messages" : [
    "Attempted to authenticate user Jane",
    "Connecting to internal IAM system",
    "Internal IAM system responded with unknown user id",
    "Provided user id is unknown, cannot verify user, returning response to user",
    "…",
    "Requested completed"
  ]
}
```

There are two fields here that aren't strings, `duration` (an integer) and `messages` (an array). 

Would be great if LogQL would support querying against array fields like this.

@cyriltovena let me know what you think!

Reply:
I haven't thought about the full json syntax yet.